optuna/__init__.py

```diff
@@ -20,3 +22,29 @@
 from optuna.study import Study  # NOQA
 from optuna.trial import Trial  # NOQA
 from optuna.version import __version__  # NOQA
+from optuna.type_checking import TYPE_CHECKING  # NOQA
+
+
+if TYPE_CHECKING:
+    from optuna import dashboard  # NOQA
+    from typing import Any  # NOQA
+else:
+
+    class _LazyImport(types.ModuleType):
+        def __init__(self, name):
+            # type: (str) -> None
+            super(_LazyImport, self).__init__(name)
+            self._name = name
+
+        def _load(self):
+            # type: () -> types.ModuleType
+            module = importlib.import_module(self._name)
+            globals()[self._name] = module
```

Comment:
It avoids unnecessary indirect access to the loader class.

Sample Code:
```python
class DummyModule:
    attr = 1


class IndirectLoader:
    def _load(self):
        return DummyModule()

    def __getattr__(self, item):
        print("indirect access 1")
        return getattr(self._load(), item)


class DirectLoader:
    def _load(self):
        m = DummyModule()
        globals()['module2'] = m
        return m

    def __getattr__(self, item):
        print("indirect access 2")
        return getattr(self._load(), item)


print("test indirect")
module1 = IndirectLoader()
module1.attr
module1.attr

print("test direct")
module2 = DirectLoader()
module2.attr
module2.attr
```

Ouput:
```
test indirect
indirect access 1
indirect access 1
test direct
indirect access 2
```

Reply:
Thank you for the explanation. I understood that reason.
But, personally, I prefer to use `setattr` function for that purpose (such as [the lazy loading of integration modules](https://github.com/optuna/optuna/blob/master/optuna/integration/__init__.py#L76)) because it only affects the local state of the loader class (I would like to avoid mutating global states unless it's really necessary).
```python
# An example usage of `setattr`.
class DirectLoader2:
    def _load(self):
        return DummyModule()

    def __getattr__(self, item):
        print("indirect access 3")
        value = getattr(self._load(), item)
        setattr(self, item, value)
        return value

# print("test direct2")
# module3 = DirectLoader2()
# module3.attr
# module3.attr
# 
# Output:
#   test direct2
#   indirect access 3
```

Reply:
We discussed this offline and decided to remove the use of `globals` function.